File: clients/CSharp/Model/AttackAction.cs
namespace Aicup2020.Model
{
    public struct AttackAction
    {
        public int? Target { get; set; }
        public Model.AutoAttack? AutoAttack { get; set; }
        public AttackAction(int? target, Model.AutoAttack? autoAttack)
        {
            this.Target = target;
            this.AutoAttack = autoAttack;
        }
        public static AttackAction ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new AttackAction();
            if (reader.ReadBoolean())
            {
                result.Target = reader.ReadInt32();
            } else
            {
                result.Target = null;
            }
            if (reader.ReadBoolean())
            {
                result.AutoAttack = Model.AutoAttack.ReadFrom(reader);
            } else
            {
                result.AutoAttack = null;
            }
            return result;
        }
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            if (!Target.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                writer.Write(Target.Value);
            }
            if (!AutoAttack.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                AutoAttack.Value.WriteTo(writer);
            }
        }
    }
}

File: clients/FSharp/Model/AttackAction.fs
#nowarn "0058"
namespace Aicup2020.Model
type AttackAction = {
    Target: option<int>;
    AutoAttack: option<AutoAttack>;
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this.Target with
            | Some value ->
                writer.Write true
                writer.Write value
            | None -> writer.Write false
        match this.AutoAttack with
            | Some value ->
                writer.Write true
                value.writeTo writer
            | None -> writer.Write false
    static member readFrom(reader: System.IO.BinaryReader) = {
        Target = match reader.ReadBoolean() with
            | true ->
                Some(
                    reader.ReadInt32()
                    )
            | false -> None
        AutoAttack = match reader.ReadBoolean() with
            | true ->
                Some(
                    AutoAttack.readFrom reader
                    )
            | false -> None
    }
